bindings/mqtt/mqtt.go

```diff
@@ -6,114 +6,258 @@
+		var err error
+		m.cleanSession, err = strconv.ParseBool(val)
+		if err != nil {
+			return &m, fmt.Errorf("%s invalid clean session %s, %s", errorMsgPrefix, val, err)
+		}
+	}
+
+	if val, ok := md.Properties[mqttCACert]; ok && val != "" {
+		if !isValidPEM(val) {
+			return &m, fmt.Errorf("%s invalid ca certificate", errorMsgPrefix)
+		}
+		m.tlsCfg.caCert = val
+	}
+	if val, ok := md.Properties[mqttClientCert]; ok && val != "" {
+		if !isValidPEM(val) {
+			return &m, fmt.Errorf("%s invalid client certificate", errorMsgPrefix)
+		}
+		m.tlsCfg.clientCert = val
+	}
+	if val, ok := md.Properties[mqttClientKey]; ok && val != "" {
+		if !isValidPEM(val) {
+			return &m, fmt.Errorf("%s invalid client certificate key", errorMsgPrefix)
+		}
+		m.tlsCfg.clientKey = val
+	}
+
+	if val, ok := md.Properties[mqttBackOffMaxRetries]; ok && val != "" {
+		backOffMaxRetriesInt, err := strconv.Atoi(val)
+		if err != nil {
+			return &m, fmt.Errorf("%s invalid backOffMaxRetries %s, %s", errorMsgPrefix, val, err)
+		}
+		m.backOffMaxRetries = backOffMaxRetriesInt
+	}
+
+	return &m, nil
 }
 
-func (m *MQTT) getMQTTMetadata(metadata bindings.Metadata) (*mqttMetadata, error) {
-	b, err := json.Marshal(metadata.Properties)
+// Init does MQTT connection parsing
+func (m *MQTT) Init(metadata bindings.Metadata) error {
+	mqttMeta, err := parseMQTTMetaData(metadata)
 	if err != nil {
-		return nil, err
+		return err
 	}
+	m.metadata = mqttMeta
 
-	var mMetadata mqttMetadata
-	err = json.Unmarshal(b, &mMetadata)
+	// mqtt broker allows only one connection at a given time from a clientID.
+	producerClientID := fmt.Sprintf("%s-producer", m.metadata.clientID)
+	p, err := m.connect(producerClientID)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	return &mMetadata, nil
+	m.ctx, m.cancel = context.WithCancel(context.Background())
+
+	// TODO: Make the backoff configurable for constant or exponential
+	b := backoff.NewConstantBackOff(5 * time.Second)
+	m.backOff = backoff.WithContext(b, m.ctx)
+
+	m.producer = p
+
+	m.logger.Debug("mqtt message bus initialization complete")
+
+	return nil
 }
 
 func (m *MQTT) Operations() []bindings.OperationKind {
 	return []bindings.OperationKind{bindings.CreateOperation}
 }
 
 func (m *MQTT) Invoke(req *bindings.InvokeRequest) (*bindings.InvokeResponse, error) {
-	m.client.Publish(m.metadata.Topic, 0, false, string(req.Data))
-	m.client.Disconnect(0)
+	m.logger.Debugf("mqtt publishing topic %s with data: %v", m.metadata.topic, req.Data)
+
+	token := m.producer.Publish(m.metadata.topic, m.metadata.qos, m.metadata.retain, req.Data)
+	if !token.WaitTimeout(defaultWait) || token.Error() != nil {
+		return nil, fmt.Errorf("mqtt error from publish: %v", token.Error())
+	}
 
 	return nil, nil
 }
 
 func (m *MQTT) Read(handler func(*bindings.ReadResponse) ([]byte, error)) error {
-	c := make(chan os.Signal, 1)
-	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+	sigterm := make(chan os.Signal, 1)
```

Comment:
I can see why you did this. The binding interfaces to not contain a `Close() error` method or embed `io.Closer`. I'm just making a comment here for tracking purposes. I opened issue #779 to make this cleaner.